docs/static/persistent-queues.asciidoc

```diff
@@ -1,44 +1,197 @@
+`queue.checkpoint.writes`:: Sets the maximum number of written events before a forced checkpoint. 
+Default is `1024`. Set to `0` for unlimited.
++
+To avoid losing data in the persistent queue, you can set `queue.checkpoint.writes: 1` to force a checkpoint after each event is
+written. Keep in mind that disk writes have a resource cost. Setting this value
+to `1` ensures maximum durability, but can severely impact performance.
+See <<durability-persistent-queues>> to better understand the trade-offs.
+`queue.checkpoint.interval`:: Sets the interval in milliseconds when a checkpoint is forced on the head page.
+Default is `1000`. Set to `0` to eliminate periodic checkpoints.
+
+[[pq-config-notes]]
+==== Configuration notes
+
+Every situation and environment is different, and the "ideal" configuration varies.
+If you optimize for performance, you may increase your risk of losing data.
+If you optimize for data protection, you may impact performance. 
+
+[[pq-size]]
+===== Queue size
+You can control queue size with the `queue.max_events` and  `queue.max_bytes` settings.
+If both settings are specified, Logstash uses whichever criteria is reached
+first. 
+See <<backpressure-persistent-queue>> for behavior when queue limits are
+reached.
+
+[[pq-lower-max_bytes]]
+===== Smaller queue size
+If you are using persistent queues to protect against data loss, but don't
+require much buffering, you can set `queue.max_bytes` to a smaller value, such
+as 10mb.
+A smaller value produces smaller queues and improves queue performance.
+
+*Sample configuration*
+
+[source, yaml]
+-----
+queue.type: persisted
+queue.max_bytes: 4gb 
+-----
+
+[[pq-fewer-checkpoints]]
+===== Fewer checkpoints
+
+Setting `queue.checkpoint.writes` and `queue.checkpoint.acks` to `0` may
+yield maximum performance, but may have potential impact on durability.
+
+In a situation where Logstash is terminated or there is a hardware-level
+failure, any data that has not been checkpointed, is lost. 
+See <<durability-persistent-queues>> to better understand the trade-offs.
+
+[[troubleshooting-pqs]]
+==== Troubleshooting persistent queues
+
+If you see an error message similar to this one, you likely have a persistent queue problem.
+
+```
+message=>"java.io.IOException: Page file size is too small to hold elements
+```
+
+This error indicates that the head page (the oldest in a folder and the one with lowest page id) has a size < 18 bytes, the size of a page header.
+
+
+To resolve the issue: 
 
-The following are problems not solved by the persistent queue feature:
+. Identify the queue (or queues) that may be corrupt.
+. Stop Logstash.
+. Run `pqrepair <path>` for each of the corrupted queues.
+
+[[pqcheck]]
+===== `pqcheck` utility
+
+Run the `pqcheck` utility to identify which persistent queue--or queues--have been corrupted. 
+
+////
+//Info to add:
+* run from where? Can you run at the folder level (or pass a path) and have it search multiple queues at once? 
```

Comment:
Running from LOGSTASH_HOME, you'd do something like

```
bin/pqcheck <QUEUE_FOLDER>
```

where `QUEUE_FOLDER` is the fully qualified path to the location of the PQ

Sample run of `--help`: 

```
ubuntu@bigger:/usr/share/logstash$ bin/pqcheck --help
Using bundled JDK: /usr/share/logstash/jdk
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
usage: pqcheck [PQ dir path]
  default [PQ dir path] is data/queue/main
```

Sample in good shape:
```
ubuntu@bigger:/usr/share/logstash$ bin/pqcheck /var/lib/logstash/queue/main/
Using bundled JDK: /usr/share/logstash/jdk
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Checking queue dir: /var/lib/logstash/queue/main
checkpoint.1, fully-acked: NO, page.1 size: 67108864
pageNum=1, firstUnackedPageNum=0, firstUnackedSeqNum=239675, minSeqNum=239675, elementCount=218241, isFullyAcked=no
checkpoint.head, fully-acked: NO, page.2 size: 67108864
pageNum=2, firstUnackedPageNum=0, firstUnackedSeqNum=457916, minSeqNum=457916, elementCount=11805, isFullyAcked=no
checkpoint.0, fully-acked: NO, page.0 size: 67108864
pageNum=0, firstUnackedPageNum=0, firstUnackedSeqNum=176126, minSeqNum=1, elementCount=239674, isFullyAcked=no
```

Sample with a corrupted page file:
```
bin/pqcheck /var/lib/logstash/queue/main/
Using bundled JDK: /usr/share/logstash/jdk
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Checking queue dir: /var/lib/logstash/queue/main
checkpoint.head, fully-acked: NO, page.2 size: NOT FOUND
pageNum=2, firstUnackedPageNum=2, firstUnackedSeqNum=534041, minSeqNum=457916, elementCount=76127, isFullyAcked=no
```

Reply:
Note that `NOT FOUND` in there - that was due to a corrupted page file

Reply:
Checkpoints may show `fully-acked: NO`, or `fully-acked: YES`. Once all the elements in the page file have been read (events ingested), the status of `fully-acked` will go from `NO` to `YES`

Reply:
The other numbers refer to the items inside each of the page files, what are the message sequence numbers, how many elements are in the page, and how far has Logstash progressed through the page file.